src/Stratis.Bitcoin/P2P/Peer/NetworkPeerClient.cs

```diff
@@ -145,6 +154,148 @@ public NetworkPeerClient(int Id, TcpClient tcpClient, ILoggerFactory loggerFacto
+        /// <exception cref="OperationCanceledException">Thrown if the operation was cancelled or the end of the stream was reached.</exception>
+        /// <remarks>
+        /// Each networkm message starts with the magic value. If the connection stream is in unknown state,
+        /// the next bytes to read might not be the magic. Therefore we read from the stream until we find the magic value.
+        /// </remarks>
+        public async Task ReadMagicAsync(byte[] magic, CancellationToken cancellation)
+        {
+            this.logger.LogTrace("()");
+
+            byte[] bytes = new byte[1];
+            for (int i = 0; i < magic.Length; i++)
+            {
+                byte expectedByte = magic[i];
+
+                await this.ReadBytesAsync(bytes, 0, bytes.Length, cancellation).ConfigureAwait(false);
+
+                byte receivedByte = bytes[0];
+                if (expectedByte != receivedByte)
+                {
+                    // If we did not receive the next byte we expected
+                    // we either received the first byte of the magic value
+                    // or not. If yes, we set index to 0 here, which is then
+                    // incremented in for loop to 1 and we thus continue 
+                    // with the second byte. Otherwise, we set index to -1 
+                    // here, which means that after the loop incrementation,
+                    // we will start from first byte of magic.
+                    i = receivedByte == magic[0] ? 0 : -1;
+                }
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Reads a specific number of bytes from the connection stream into a buffer.
+        /// </summary>
+        /// <param name="buffer">Buffer to read incoming data to.</param>
+        /// <param name="offset">Position in the buffer where to write the data.</param>
+        /// <param name="bytesToRead">Number of bytes to read.</param>
+        /// <param name="cancellation">Cancellation token that allows aborting the read operation.</param>
+        /// <returns>Binary data received from the connected counterparty.</returns>
+        /// <exception cref="OperationCanceledException">Thrown if the operation was cancelled or the end of the stream was reached.</exception>
+        private async Task ReadBytesAsync(byte[] buffer, int offset, int bytesToRead, CancellationToken cancellation = default(CancellationToken))
+        {
+            this.logger.LogTrace("({0}:{1},{2}:{3})", nameof(offset), offset, nameof(bytesToRead), bytesToRead);
+
+            while (bytesToRead > 0)
+            {
+                int chunkSize = await this.Stream.ReadAsync(buffer, offset, bytesToRead, cancellation).ConfigureAwait(false);
+                if (chunkSize == 0)
+                {
+                    this.logger.LogTrace("(-)[STREAM_END]");
+                    throw new OperationCanceledException();
```

Comment:
This `OperationCanceledException` will disconnect the node no?

Reply:
Ah or is that your intention..

Reply:
yes